lib/elixir/lib/kernel/typespec.ex

```diff
@@ -535,10 +535,29 @@ defmodule Kernel.Typespec do
   end
 
   # Handle type operator
-  defp typespec({:::, meta, [var, expr]}, vars, caller) do
-    left = typespec(var, [elem(var, 0) | vars], caller)
-    right = typespec(expr, vars, caller)
-    {:ann_type, line(meta), [left, right]}
+  defp typespec({:::, meta, [var, expr]} = orig, vars, caller) do
+    case var do
+      {var_name, _meta, context} when is_atom(var_name) and is_atom(context) ->
+        case typespec(expr, vars, caller) do
+          {:ann_type, _, _} ->
```

Comment:
I'm not 100% sure if this error should be raised, but I couldn't think if any valid use case for nested type annotations. 

Besides, I will raise a `CompileError` with typing errors like this one `@type my_type, {pid, name :: atom :: atom}`🙂

Thoughts?